content/Improve your Rails App searches with Elasticsearch - Leandro Dominguez.md

```diff
@@ -0,0 +1,214 @@
+  },
+  "tagline" : "You Know, for Search"
+}
+```
+
+Now we have our Elastic instance up and running, let's open a Rails console and create the corresponding index by executing:
+
+```ruby
+Post.__elasticsearch__.create_index!
+```
+
+If you already had some posts created, you will need to reindex them by running the following:
+
+```ruby
+Post.import
+```
+
+This can take A LOT of time if you have too many records, since Elastic has to analyze and tokenize all of them.
+
+**Step 5: define the search method we will call from the controller**
+
+In the Posts’ model we'll define this basic method that will execute a query in our Elasticsearch instance, taking a string and looking in the specified fields.
+
+```ruby
+   def self.our_first_query(query)
+     search(
+       query: {
+         multi_match: {
+           query: query,
+           fields: %w[title body^5]
+         }
+       }
+     )
+   end
+```
+
+This simple query will search through all our posts and retrieve them in order of importance. The ‘`^5`’ after `body` will make a match in this attribute to weigh 5 more times than one on `title`.
+
+**Step 6: create your searches method in the controller, its view and routes.**
+
+My controller method looks just like this
+
+```ruby
+ def search
+   return unless query.present?
+
+   @posts = Post.our_first_query(query)
+ end
+```
+
+Also, I did this simple view with a little of Bootstrap.
+
+![Search by semantic information](images/semantic_search.png)
+
+As you can see, I created a bunch of posts using [faker]([https://github.com/faker-ruby/faker](https://github.com/faker-ruby/faker)).
+
+Take a look into that search again… Don’t you see something special? We looked for ‘_darkness_’ and although we don’t have any matching result, we can see that the ones containing ‘_dark_’ were displayed. This is because of the text analyzers. They know these two words are not the same but they do have similar semantics.
+
+Pretty easy, wasn’t it?
+
+What happens if we look for ‘_dakr_’ instead of ‘_dark_’?
+
+![No results found](images/with_typos.png)
+
+Oops, no results. Let’s fix this.
+
+**Step 7: searching with typos and highlighting the match in searches’ results.**
+
+Our search will now look like this:
+
+```ruby
+def self.highlight(query)
+ search(
+   query: {
+     multi_match: {
+       query: query,
+       fuzziness: 'AUTO',
+       fields: %w[title body]
+     }
+   },
+   highlight: {
+     pre_tags: ['<b>'],
+     post_tags: ['</b>'],
+     fields: {
+       body: {}
+     }
+   }
+ )
+end
+```
+
+In this case we are basically adding new two features: _fuzziness_ and _highlight_.
+
+With fuzziness we are making Elastic know that errors may appear in the query. This way, it will try to match similar words, so we now will not only cover semantic but syntactic errors too.
+
+With the second part, we are making the match to be highlighted by adding pre and post tags (in this case HTML bold, but you can add whatever you want).
```

Comment:
```suggestion
With highlight, we are making the match to be highlighted by adding pre and post tags (in this case HTML bold, but you can add whatever you want).
```

I think it's better if you just say what was the second part even though you repeat it, like you did in the previous paragraph.